packaging/requirements.py

```diff
@@ -144,3 +144,9 @@ def __str__(self) -> str:
 
     def __repr__(self) -> str:
         return f"<Requirement('{self}')>"
+
```

Comment:
What does this have to do with metadata?

Reply:
One nice side effect of having metadata as a fully immutable/frozen data structure is that we can have nice things like comparing fields.

So for example someone could do:
```python
metadata1.provides_dist == metadata2.provides_dist
# or even
metadata1 == metadata2
```

`requires_dist`, `provides_dist` and `obsolete_dist` all are currently all implemented as frozen sets of `Requirement`. While a `frozenset` itself is comparable, the comparison will only work properly as long as all the members are hashable themselves (please correct me if I am wrong here).

By making `Requirements` hashable this "nice-to-have" property on metadata is achieved.

While I like this feature, I don't mind removing it.
However the current test implementation heavily rely on this property and will fail if `Requirement` are not hashable, so if this change is not welcome, please let me know and I will think a bit more and rewrite the tests.

Reply:
It might be worth submitting this separately.

Reply:
No problems.

I have updated the implementation without making the `Requirement` class hashable and changed the tests accordingly.

I also prevented `dataclass` from generating any comparison methods in `CoreMetadata` to avoid misleading outcomes.